src/dataflow/src/render/sources.rs

```diff
@@ -816,6 +862,129 @@ where
+/// NOTE: There can only ever be one running (rendered) instance of a persisted source. The manager
+/// will panic when trying to add more than on instance for a given source id.
+pub struct PersistedSourceManager {
+    /// Handles that allow setting the compaction frontier for a persisted source.
+    compaction_handles: HashMap<GlobalId, Weak<BoundedCompactionHandle>>,
+}
+
+impl PersistedSourceManager {
+    pub fn new() -> Self {
+        PersistedSourceManager {
+            compaction_handles: HashMap::new(),
+        }
+    }
+
+    /// Binds the given `frontier_probe` and `compaction_handle` to the `source_id`. This will
+    /// panic if there is also ready a source for the given id. This returns a "token" that should
+    /// be stashed in the customary places to prevent the stored information from vanishing before
+    /// its time.
+    pub fn add_source(
+        &mut self,
+        source_id: &GlobalId,
+        source_name: &str,
+        frontier_probe: ProbeHandle<Timestamp>,
+        compaction_handle: MultiWriteHandle,
+    ) -> Rc<dyn Any> {
+        let compaction_handle = BoundedCompactionHandle::new(frontier_probe, compaction_handle);
+
+        let compaction_token = Rc::new(compaction_handle);
+
+        let previous = self
+            .compaction_handles
+            .insert(source_id.clone(), Rc::downgrade(&compaction_token));
+
+        match previous {
+            Some(compaction_handle) => {
+                if compaction_handle.upgrade().is_some() {
+                    panic!(
+                        "cannot have multiple rendered instances for persisted source {} ({})",
+                        source_id, source_name
+                    );
+                }
+            }
+            None => (), // All good!
+        }
+
+        compaction_token as Rc<dyn Any>
+    }
+
+    /// Enables compaction of sources associated with the identifier.
+    pub fn allow_compaction(&mut self, id: GlobalId, frontier: Antichain<Timestamp>) {
+        if let Some(compaction_handle) = self.compaction_handles.get(&id) {
+            let compaction_handle = match compaction_handle.upgrade() {
+                Some(handle) => handle,
+                None => {
+                    // Clean up the expired weak handle.
+                    //
+                    // NOTE: Right now, there is no explicit message when a rendered source
+                    // instance is being dropped .Sources are being droppped implicitly when all
+                    // their consumers are being dropped, and we notice that when the token that
+                    // holds on to the strong Rc to the compaction handle is being dropped.
+                    self.compaction_handles.remove(&id);
+                    return;
+                }
+            };
+
+            compaction_handle.allow_compaction(frontier);
+        }
+    }
+
+    /// Removes the maintained state for source `id`.
+    pub fn del_source(&mut self, id: &GlobalId) -> bool {
```

Comment:
very very minor nit: can we call this `delete_source` or `drop_source`?

Reply:
I took the name from `TraceManager`, but yes, it's a bit weird. Happy to change it, but then the names diverge.